### ISSUE TITLE
❓ why always n_layers >=2 in gnn models 

### ISSUE THREAD

**Issue by caowencai** [dmlc/dgl#2269]
## ❓ why always n_layers >=2 in gnn models like  graphsage
I have found that many models implemented in dgl have at least two layers, and featute dimension changes through: in_feats -> n_hidden -> n_classes.
Then why there always exists at least two layers in GNN models inherently (and with nn.linear included), for better performance? 
Why not implement end-to-end task  from dimension  in_feats to n_classes with only one  GNN layer?

Here is the graphsage code:
```python
class SAGE(nn.Module):
    def __init__(self,
                 in_feats,
                 n_hidden,
                 n_classes,
                 n_layers,
                 activation,
                 dropout):
        super().__init__()
        self.n_layers = n_layers
        self.n_hidden = n_hidden
        self.n_classes = n_classes
        self.layers = nn.ModuleList()
        self.layers.append(dglnn.SAGEConv(in_feats, n_hidden, 'mean'))
        for i in range(1, n_layers - 1):
            self.layers.append(dglnn.SAGEConv(n_hidden, n_hidden, 'mean'))
        self.layers.append(dglnn.SAGEConv(n_hidden, n_classes, 'mean'))
        self.dropout = nn.Dropout(dropout)
        self.activation = activation

**Comment by mufeili** [comment]
An implicit assumption here is that applying a non-linearity after the aggregation result makes the model more expressive. With a single GNN layer, you cannot apply an activation layer after the aggregation result in cases like node classification.

**Comment by caowencai** [comment]
@mufeili  Got it, thanks.

**Comment by caowencai** [comment]
@mufeili 
So is there any more assumption that the second GNN is better than nn.linear individually as out layer?

**Comment by mufeili** [comment]
> @mufeili
> So is there any more assumption that the second GNN is better than nn.linear individually as out layer?

The number of GNN layers to use is related to the number of hops that matters for message passing.

**Comment by caowencai** [comment]
I see.
